src/cloud.c

```diff
@@ -426,88 +418,102 @@ int cloud_publish_data(const char *id, uint8_t sensor_id, uint8_t value_type,
 		       const knot_value_type *value,
 		       uint8_t kval_len)
 {
-	amqp_bytes_t queue_cloud;
 	json_object *jobj_data;
 	const char *json_str;
 	int result;
 
-	queue_cloud = mq_declare_new_queue(MQ_QUEUE_FOG_IN);
-	if (!queue_cloud.bytes) {
-		l_error("Error on declare a new queue");
-		return -1;
-	}
-
 	jobj_data = parser_data_create_object(id, sensor_id, value_type, value,
 					      kval_len);
-	if (!jobj_data) {
-		amqp_bytes_free(queue_cloud);
+	if (!jobj_data)
 		return KNOT_ERR_CLOUD_FAILURE;
-	}
+
 	json_str = json_object_to_json_string(jobj_data);
 
 	headers[0].value.value.bytes = amqp_cstring_bytes(user_auth_token);
 
-	result = mq_publish_persistent_message(queue_cloud,
-					       MQ_EXCHANGE_FOG_IN,
-					       MQ_CMD_DATA_PUBLISH,
-					       headers, 1,
-					       MQ_MSG_EXPIRATION_TIME_MS,
-					       json_str);
+	result = mq_publish_fanout_message(MQ_EXCHANGE_DATA_SENT,
+					   headers, 1,
+					   MQ_MSG_EXPIRATION_TIME_MS,
+					   json_str);
 	if (result < 0)
 		result = KNOT_ERR_CLOUD_FAILURE;
 
 	json_object_put(jobj_data);
-	amqp_bytes_free(queue_cloud);
+
 	return result;
 }
 
 /**
  * cloud_set_read_handler:
+ * @id: thing ids
  * @cb: callback to handle message received from cloud
  * @user_data: user data provided to callbacks
  *
  * Set callback handler when receive cloud messages.
  *
  * Returns: 0 if successful and -1 otherwise.
  */
-int cloud_set_read_handler(cloud_cb_t read_handler, void *user_data)
+int cloud_set_read_handler(const char *id, cloud_cb_t read_handler,
+			   void *user_data)
 {
-	const char *fog_events[] = {
-		MQ_EVENT_DATA_UPDATE,
-		MQ_EVENT_DATA_REQUEST,
-		MQ_EVENT_DEVICE_REGISTERED,
-		MQ_EVENT_DEVICE_UNREGISTERED,
-		MQ_EVENT_DEVICE_AUTH,
-		MQ_EVENT_SCHEMA_UPDATED,
-		NULL
-	};
 	amqp_bytes_t queue_fog;
-	int err, i;
+	char queue_fog_name[100];
+	char queue_reply_name[100];
+	int msg_type;
+	int err;
 
 	cloud_cb = read_handler;
 
-	queue_fog = mq_declare_new_queue(MQ_QUEUE_FOG_OUT);
+	if (set_cloud_events(id))
+		return -1;
+
+	snprintf(queue_fog_name, sizeof(queue_fog_name), "%s-%s",
+		 MQ_QUEUE_FOG_OUT, id);
+
+	queue_fog = mq_declare_new_queue(queue_fog_name);
 	if (queue_fog.bytes == NULL) {
 		l_error("Error on declare a new queue");
 		return -1;
 	}
 
-	for (i = 0; fog_events[i] != NULL; i++) {
-		err = mq_bind_queue(queue_fog, MQ_EXCHANGE_FOG_OUT, fog_events[i]);
+	for (msg_type = UPDATE_MSG; msg_type < MSG_TYPES_LENGTH; msg_type++) {
+		err = mq_prepare_direct_queue(queue_fog, MQ_EXCHANGE_DEVICE,
+					      cloud_events[msg_type]);
 		if (err) {
 			l_error("Error on set up queue to consume");
 			amqp_bytes_free(queue_fog);
 			return -1;
 		}
 	}
 
-	err = mq_set_read_cb(queue_fog, on_cloud_receive_message, user_data);
+	err = mq_set_read_cb(on_cloud_receive_message, user_data);
 	if (err) {
 		l_error("Error on set up read callback");
 		return -1;
 	}
+
+	err = mq_consumer_queue(queue_fog);
+	if (err) {
+		l_error("Error on start a queue consumer\n");
```

Comment:
Is the '\n' character necessary to improve the log readability?

Reply:
Done